{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"mount_file_id":"1kWvgagnSecBVVKgEkrRMW9elobZfhvNg","authorship_tag":"ABX9TyNLm9qLNrzWjGfvJA5+UA5u"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["import pandas\n","import numpy\n","import keras\n","import matplotlib.pyplot as plt\n","from matplotlib.image import imread"],"metadata":{"id":"j9ceEGxEKZMx"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["path = '/content/drive/MyDrive/Colab Notebooks/AI/Chessman/archive/Chessman-image-dataset/Chess'\n","img_height,img_width,batch_size,epochs = 224,224,64,10\n","train_df = keras.utils.image_dataset_from_directory(\n","    path,\n","    validation_split = 0.1,\n","    subset = 'training',\n","    batch_size = batch_size,\n","    image_size = (img_height, img_width),\n","    shuffle=True,\n","    seed = 123,\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Acf8LhwvKuRQ","executionInfo":{"status":"ok","timestamp":1722318640121,"user_tz":-330,"elapsed":604,"user":{"displayName":"Shreesha Study","userId":"13095551360219515715"}},"outputId":"9539c7a5-e860-49e4-ed34-02fe23329fc0"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Found 552 files belonging to 6 classes.\n","Using 497 files for training.\n"]}]},{"cell_type":"code","source":["label_df = keras.utils.image_dataset_from_directory(\n","    path,\n","    validation_split = 0.2,\n","    subset = 'validation',\n","    batch_size = batch_size,\n","    image_size = (img_height, img_width),\n","    shuffle=True,\n","    seed = 123,\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SSCwNUqsMOHJ","executionInfo":{"status":"ok","timestamp":1722318720733,"user_tz":-330,"elapsed":400,"user":{"displayName":"Shreesha Study","userId":"13095551360219515715"}},"outputId":"5ff90099-d893-4209-e39b-b431eb9cbd00"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Found 552 files belonging to 6 classes.\n","Using 110 files for validation.\n"]}]},{"cell_type":"code","source":["num_classes = len(train_df.class_names)"],"metadata":{"id":"-JJfUL0fNWx1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["model = keras.models.Sequential()\n","model.add(keras.layers.Conv2D(8, (3, 3), padding=\"same\",input_shape=(224,224,3), activation=\"relu\"))\n","model.add(keras.layers.MaxPooling2D(pool_size=(3, 3)))\n","model.add(keras.layers.Conv2D(16, (3, 3), padding=\"same\", activation=\"relu\"))\n","model.add(keras.layers.MaxPooling2D(pool_size=(2, 2)))\n","model.add(keras.layers.Conv2D(32, (3, 3), padding=\"same\", activation=\"relu\"))\n","model.add(keras.layers.MaxPooling2D(pool_size=(2, 2)))\n","model.add(keras.layers.Flatten())\n","model.add(keras.layers.Dense(32, activation=\"relu\"))\n","model.add(keras.layers.Dense(num_classes, activation=\"softmax\"))\n","model.summary()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"sOzMVn9tMWmh","executionInfo":{"status":"ok","timestamp":1722319050958,"user_tz":-330,"elapsed":460,"user":{"displayName":"Shreesha Study","userId":"13095551360219515715"}},"outputId":"cb0a8470-95e6-458b-87c8-49f90379a701"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Model: \"sequential\"\n","_________________________________________________________________\n"," Layer (type)                Output Shape              Param #   \n","=================================================================\n"," conv2d (Conv2D)             (None, 224, 224, 8)       224       \n","                                                                 \n"," max_pooling2d (MaxPooling2  (None, 74, 74, 8)         0         \n"," D)                                                              \n","                                                                 \n"," conv2d_1 (Conv2D)           (None, 74, 74, 16)        1168      \n","                                                                 \n"," max_pooling2d_1 (MaxPoolin  (None, 37, 37, 16)        0         \n"," g2D)                                                            \n","                                                                 \n"," conv2d_2 (Conv2D)           (None, 37, 37, 32)        4640      \n","                                                                 \n"," max_pooling2d_2 (MaxPoolin  (None, 18, 18, 32)        0         \n"," g2D)                                                            \n","                                                                 \n"," flatten (Flatten)           (None, 10368)             0         \n","                                                                 \n"," dense (Dense)               (None, 32)                331808    \n","                                                                 \n"," dense_1 (Dense)             (None, 6)                 198       \n","                                                                 \n","=================================================================\n","Total params: 338038 (1.29 MB)\n","Trainable params: 338038 (1.29 MB)\n","Non-trainable params: 0 (0.00 Byte)\n","_________________________________________________________________\n"]}]},{"cell_type":"code","source":["model.compile(loss = 'categorical_crossentropy', optimizer = keras.optimizers.Adam(0.0005),metrics=['accuracy'])"],"metadata":{"id":"FybKYYODNyIH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["history = model.fit(train_df, validation_data=label_df, epochs=10)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":703},"id":"YY2Y996VOqzl","executionInfo":{"status":"error","timestamp":1722319324734,"user_tz":-330,"elapsed":445,"user":{"displayName":"Shreesha Study","userId":"13095551360219515715"}},"outputId":"6b29a502-b94f-4785-d82a-7b703271955b"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["Epoch 1/10\n"]},{"output_type":"error","ename":"ValueError","evalue":"in user code:\n\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function  *\n        return step_function(self, iterator)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step  **\n        outputs = model.train_step(data)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1151, in train_step\n        loss = self.compute_loss(x, y, y_pred, sample_weight)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1209, in compute_loss\n        return self.compiled_loss(\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 277, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 143, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 270, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 2221, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend.py\", line 5573, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (None, 1) and (None, 6) are incompatible\n","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m<ipython-input-17-018a29e03a31>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhistory\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabel_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     68\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m             \u001b[0;31m# `tf.debugging.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\u001b[0m in \u001b[0;36mtf__train_function\u001b[0;34m(iterator)\u001b[0m\n\u001b[1;32m     13\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m                     \u001b[0mdo_return\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m                     \u001b[0mretval_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconverted_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstep_function\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mag__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mld\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfscope\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m                 \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m                     \u001b[0mdo_return\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: in user code:\n\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function  *\n        return step_function(self, iterator)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step  **\n        outputs = model.train_step(data)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1151, in train_step\n        loss = self.compute_loss(x, y, y_pred, sample_weight)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1209, in compute_loss\n        return self.compiled_loss(\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 277, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 143, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 270, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 2221, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend.py\", line 5573, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (None, 1) and (None, 6) are incompatible\n"]}]}]}